### DISCUSSION TITLE
Scheme methods, .NET delegates

### DISCUSSION THREAD

**Discussion by jtoverka** [leppie/IronScheme#125] [Discussions] [answered]
@leppie 

The code below works. However, there is commented code at the bottom that does not work. How do I assign a method created in Scheme to a delegate? For the sake of simplicity, I want to assign the msg method in scheme to a .NET delegate called System.Action. How do I go about this?

```
;; (command "-SCMLOAD" "C:/Users/Anonymous/OneDrive/Documents/test.scm")
(interaction-environment (new-interaction-environment))
(import (rnrs) (ironscheme clr))

(clr-reference "C:\\Program Files\\Autodesk\\AutoCAD 2023\\acdbmgd.dll")
(clr-reference "C:\\Program Files\\Autodesk\\AutoCAD 2023\\acmgd.dll")
(clr-reference "C:\\Program Files\\Autodesk\\AutoCAD 2023\\accoremgd.dll")

(clr-using Autodesk.AutoCAD.ApplicationServices.Core)
(clr-using Autodesk.AutoCAD.ApplicationServices)
(clr-using Autodesk.AutoCAD.EditorInput)
(clr-using Autodesk.AutoCAD.Internal)
(clr-using System)

(define msg (lambda ()
	(define documentManager (clr-static-prop-get Application DocumentManager))
	(define activeDocument (clr-prop-get DocumentCollection MdiActiveDocument documentManager))
	(define editor (clr-prop-get Document Editor activeDocument))
	(clr-call Editor WriteMessage editor "\nOur test command works!")
))

(msg)

;; (define commandCallback (clr-new Action msg))
;; (clr-static-call Utils AddCommand "scmcmds" name name 0 commandCallback)
```

**Comment by leppie** [comment]
The compiler should automatically create a delegate wrapper of the type you assign the proc to. 

So this should just work:

```
(clr-static-call Utils AddCommand "scmcmds" name name 0 msg)
```

**Reply by jtoverka** [reply]
@leppie 
I tried this and the error I get is: "attempted to use undefined symbol"

**Comment by jtoverka** [comment, marked as answer]
I have resolved this issue.
Main issue is the defined methods in Scheme are IronScheme.Runtime.Closure objects. The method (System.Reflection.MethodInfo) itself has a return type of System.Object. The delegate (Autodesk.AutoCAD.Internal.CommandCallback) needs a return type of void so there is that incompatibility there.

From C# .NET I created a function to wrap the System.Reflection.MethodInfo into a lambda expression and assign it to the delegate Autodesk.AutoCAD.Internal.CommandCallback to finally register the command.

In Scheme I just call the function I made in C# .NET to register the scheme method as an AutoCAD command.
